api/inc/register_gateway.h

```diff
@@ -17,105 +17,208 @@
-    "skip_args%=:\n"
+/** `BX LR` encoding
+ * @internal
+ */
+#define BXLR                           ((uint16_t) 0x4770)
 
-/* 2 arguments: Simple write, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA2(src_box, addr, val) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
+/** Register Gateway - Read operation
+ *
+ * This macro provides an API to perform 32-bit read operations on restricted
+ * registers. Such accesses are assembled into a read-only flash structure that
+ * is read and validated by uVisor before performing the operation.
+ *
+ * @warning These APIs currently only support link-time known values. This means
+ * that an access of this type is not supported:
+ * ```
+ * static uint32_t const const_mask = 0x0000F000;
+ * a = uvisor_read(&SYSCFG->CTRL, UVISOR_RGW_OP_READ, const_mask); // Yes
+ * uint32_t var_mask = rand();
+ * a = uvisor_read(&SYSCFG->CTRL, UVISOR_RGW_OP_READ, var_mask); // No
+ * ```
+ *
+ * @param box_name[in]  The name of the source box as decalred in
+ *                      `UVISOR_BOX_CONFIG`.
+ * @param addr[in]      The address for the data access.
+ * @param operation[in] The operation to perform at the address for the read. It
+ *                      is chosen among the `UVISOR_RGW_OP_*` macros.
+ * @param mask[in]      The mask to apply for the read operation.
+ * @returns The value read from address using the operation and mask provided
+ * (or their respective defaults if they have not been provided).
+ */
+#define uvisor_read(box_name, addr, op, msk) \
+    ({ \
+        /* Instanstiate the gateway. This gets resolved at link-time. */ \
+        static TRegisterGateway const register_gateway = { \
+            .svc_opcode = UVISOR_SVC_OPCODE(UVISOR_SVC_ID_REGISTER_GATEWAY), \
+            .branch     = BRANCH_OPCODE(OFFSETOF(TRegisterGateway, branch), OFFSETOF(TRegisterGateway, bxlr)), \
+            .magic      = UVISOR_REGISTER_GATEWAY_MAGIC, \
+            .box_ptr    = (uint32_t) & box_name ## _cfg_ptr, \
+            .address    = (uint32_t) addr, \
+            .value      = 0, \
+            .mask       = msk, \
+            .operation  = op, \
+            .bxlr       = BXLR  \
+        }; \
+        \
+        /* Pointer to the register gateway we just created. The pointer is
+         * located in a discoverable linker section. */ \
+        __attribute__((section(".keep.uvisor.register_gateway_ptr"))) \
+        static uint32_t const register_gateway_ptr = (uint32_t) &register_gateway; \
+        (void) register_gateway_ptr; \
+        \
+        /* Call the actual gateway. */ \
+        uint32_t result = ((uint32_t (*)(void)) ((uint32_t) &register_gateway | 1))(); \
```

Comment:
I'm fine with this implementation. File an issue for getting rid of this `| 1`. We can then prioritize the optimization appropriately. I wouldn't block this PR based on having the run-time `| 1` here, and it isn't a priority to get rid of it right now.